src/creature.cpp

```diff
@@ -401,32 +402,32 @@ void Creature::onCreatureMove(Creature* creature, const Tile* newTile, const Pos
 
 		if (!summons.empty()) {
 			// check if any of our summons is out of range (+/- 2 floors or 30 tiles away)
-			std::forward_list<Creature*> despawnList;
-			for (Creature* summon : summons) {
+			std::forward_list<std::shared_ptr<Creature>> despawnList;
+			for (const auto& summon : summons) {
 				const Position& pos = summon->getPosition();
 				if (newPos.getDistanceZ(pos) > 2 || std::max(newPos.getDistanceX(pos), newPos.getDistanceY(pos)) > 30) {
 					despawnList.push_front(summon);
 				}
 			}
 
-			for (Creature* despawnCreature : despawnList) {
+			for (const auto& despawnCreature : despawnList) {
 				g_game.removeCreature(despawnCreature, true);
 			}
 		}
 
 		if (newTile->getZone() != oldTile->getZone()) {
-			tfs::events::creature::onChangeZone(this, oldTile->getZone(), newTile->getZone());
+			tfs::events::creature::onChangeZone(getCreature(), oldTile->getZone(), newTile->getZone());
 			onChangeZone(getZone());
 		}
 	}
 
-	if (creature == followCreature || (creature == this && followCreature)) {
+	if (creature == followCreature || (creature.get() == this && followCreature)) {
 		if (newPos.z != oldPos.z || !canSee(followCreature->getPosition())) {
 			onCreatureDisappear(followCreature, false);
 		}
 	}
 
-	if (creature == attackedCreature || (creature == this && attackedCreature)) {
+	if (creature == attackedCreature || (creature.get() == this && attackedCreature)) {
```

Comment:
We’ve got some legacy code here, hehe!
```suggestion
	if (creature == attackedCreature || (creature == getCreature() && attackedCreature)) {
```